components/web_discovery/browser/patterns.cc

```diff
@@ -0,0 +1,321 @@
+    VLOG(1) << "Selector or key missing from payload rule";
+    return false;
+  }
+  rule_out->selector = *selector;
+  rule_out->key = *payload_key;
+  if (rule_list->size() > 2) {
+    auto* field_action = (*rule_list)[2].GetIfString();
+    if (field_action && *field_action == kJoinFieldAction) {
+      rule_out->is_join = true;
+    }
+  }
+  return true;
+}
+
+std::optional<std::vector<PayloadRuleGroup>> ParsePayloadRules(
+    const base::Value::Dict* payload_dict) {
+  std::vector<PayloadRuleGroup> result(payload_dict->size());
+
+  auto rule_group_it = result.begin();
+  for (const auto [key, rule_group_value] : *payload_dict) {
+    auto* rule_group_dict = rule_group_value.GetIfDict();
+    if (!rule_group_dict) {
+      VLOG(1) << "Payload rule group is not a dict";
+      return std::nullopt;
+    }
+    auto* action = rule_group_dict->FindString(kActionKey);
+    auto* fields = rule_group_dict->FindList(kFieldsKey);
+    auto* rule_type_str = rule_group_dict->FindString(kRuleTypeKey);
+    auto* result_type_str = rule_group_dict->FindString(kResultTypeKey);
+    if (!action || !rule_type_str || !result_type_str) {
+      VLOG(1) << "Payload rule group attributes missing";
+      return std::nullopt;
+    }
+    auto rule_type_it = kPayloadRuleTypeMap.find(*rule_type_str);
+    auto result_type_it = kPayloadResultTypeMap.find(*result_type_str);
+    if (rule_type_it == kPayloadRuleTypeMap.end() ||
+        result_type_it == kPayloadResultTypeMap.end()) {
+      VLOG(1) << "Payload rule or result types unknown";
+      return std::nullopt;
+    }
+    rule_group_it->action = *action;
+    rule_group_it->key = key;
+    rule_group_it->result_type = result_type_it->second;
+    rule_group_it->rule_type = rule_type_it->second;
+    if (fields) {
+      rule_group_it->rules = std::vector<PayloadRule>(fields->size());
+
+      auto rule_it = rule_group_it->rules.begin();
+      for (const auto& rule_value : *fields) {
+        if (!ParsePayloadRule(rule_value, rule_it.base())) {
+          return std::nullopt;
+        }
+        rule_it = std::next(rule_it);
+      }
+    }
+
+    rule_group_it = std::next(rule_group_it);
+  }
+  return result;
+}
+
+RefineFunctionList ParseFunctionsApplied(const base::Value::List* list) {
+  RefineFunctionList result;
+  for (const auto& function_val : *list) {
+    const auto* function_list = function_val.GetIfList();
+    if (!function_list || function_list->size() <= 1) {
+      continue;
+    }
+    std::vector<base::Value> function_vec;
+    for (const auto& element : *function_list) {
+      function_vec.push_back(element.Clone());
+    }
+    result.push_back(std::move(function_vec));
+  }
+  return result;
+}
+
+std::optional<base::flat_map<std::string, ScrapeRuleGroup>> ParseScrapeRules(
+    const base::Value::Dict* scrape_url_dict) {
+  base::flat_map<std::string, ScrapeRuleGroup> result;
+
+  for (const auto [selector, rule_group_value] : *scrape_url_dict) {
```

Comment:
`auto&` where possible.

Reply:
the iterator value is a pair of references

Reply:
> the iterator value is a pair of references

ok, then just `auto` and then move everything of out `rule_group_value`

you have these things as non-const which is good:
```
auto* sub_selector = rule_dict->FindString(kSubSelectorKey);
auto* attribute = rule_dict->FindString(kAttributeKey);
```

you just have to do this to take everything that `JSONReader` constructed for you (without unnecessary copying):
```
rule->attribute = std::move(*attribute);
if (sub_selector) {
  rule->sub_selector = std::move(*sub_selector);
}
```